src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -166,7 +166,18 @@ public static unsafe Task GetAddrInfoAsync(string hostName, bool justAddresses)
             SocketError errorCode = (SocketError)Interop.Winsock.GetAddrInfoExW(
                 hostName, null, Interop.Winsock.NS_ALL, IntPtr.Zero, &hints, &context->Result, IntPtr.Zero, &context->Overlapped, s_getAddrInfoExCallback, &context->CancelHandle);
 
-            if (errorCode != SocketError.IOPending)
+
+            if (errorCode == SocketError.TryAgain)
+            {
+                // WSATRY_AGAIN indicates possible problem with reachability according to docs.
+                // However, if servers are really unreachable, we would still get IOPending here
+                // and final result would be posted via overlapped IO.
+                // synchronous failure here may signal issue when GetAddrInfoExW does not work from
+                // impersonated context.
+                GetAddrInfoExContext.FreeContext(context);
+                return null;
+            }
+            else if (errorCode != SocketError.IOPending)
```

Comment:
Did we / could we add any tests that validate name resolution in an impersonated context?

Reply:
not easily as it would require separate account on the machine and password for it. So it is going to be problematic from security prospective. I was thinking about it and it seems like it would be best to address this as part of the "Enterprise authentication testing" project - if we ever decide to fund it. for that, we will need special setup and we will need to work out the password management. 

I could also add test conditional on environment variables (e.g. run if user & password is provided via environment) but that will not run during regular CI.

Reply:
We have tests at https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Principal.Windows/tests/WindowsIdentityImpersonatedTests.netcoreapp.cs for WindowsIdentity that do impersonation.  Can any of that be used?

Reply:
yes, I think this should work. I will explore that and create PR for master.